### PR TITLE
[SPARK-18801][SQL][FOLLOWUP] Alias the view with its child

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

This PR is a follow-up to address the comments https://github.com/apache/spark/pull/16233/files#r95669988 and https://github.com/apache/spark/pull/16233/files#r95662299.

We try to wrap the child by:
1. Generate the `queryOutput` by:
    1.1. If the query column names are defined, map the column names to attributes in the child output by name;
    1.2. Else set the child output attributes to `queryOutput`.
2. Map the `queryQutput` to view output by index, if the corresponding attributes don't match, try to up cast and alias the attribute in `queryOutput` to the attribute in the view output.
3. Add a Project over the child, with the new output generated by the previous steps.
If the view output doesn't have the same number of columns neither with the child output, nor with the query column names, throw an AnalysisException.

## How was this patch tested?

Add new test cases in `SQLViewSuite`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 1]
```diff
@@ -29,40 +29,31 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
-        Alias(Cast(originAttr, attr.dataType), attr.name)(exprId = attr.exprId,
-          qualifier = attr.qualifier, explicitMetadata = Some(attr.metadata))
+      if (output.length != child.output.length) {
+        throw new AnalysisException(
+          s"The view output ${output.mkString("[", ",", "]")} doesn't have the same number of " +
+            s"columns with the child output ${child.output.mkString("[", ",", "]")}")
+      }
+      val newOutput = output.zip(child.output).map {
+        case (attr, originAttr) =>
+          if (attr.dataType != originAttr.dataType) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: can you check hive's behavior? maybe we can use `UpCast` here

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: Seems that Hive supports UpCast between child output and view output, for example:
```
hive> create table testtable as select 1 a, 2L b;
hive> create view testview as select * from testtable;
hive> select * from testview;
OK
1	2
Time taken: 0.11 seconds, Fetched: 1 row(s)
hive> alter table testtable change column a a bigint;
hive> alter table testtable change column b b string;
hive> desc testtable;
OK
a                   	bigint              	                    
b                   	string              	                    
Time taken: 0.15 seconds, Fetched: 2 row(s)
hive> desc testview;
OK
a                   	int                 	                    
b                   	bigint              	                    
Time taken: 0.038 seconds, Fetched: 2 row(s)
hive> select * from testview;
OK
1	2
Time taken: 0.172 seconds, Fetched: 1 row(s)
```

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: What should we set for the `walkedTypePath` here?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: It sounds like Hive just forcefully cast it.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]:
```
hive> explain extended select * from testview;
OK
ABSTRACT SYNTAX TREE:
  
TOK_QUERY
   TOK_FROM
      TOK_TABREF
         TOK_TABNAME
            testview
   TOK_INSERT
      TOK_DESTINATION
         TOK_DIR
            TOK_TMP_FILE
      TOK_SELECT
         TOK_SELEXPR
            TOK_ALLCOLREF


STAGE DEPENDENCIES:
  Stage-0 is a root stage

STAGE PLANS:
  Stage: Stage-0
    Fetch Operator
      limit: -1
      Processor Tree:
        TableScan
          alias: testtable
          Statistics: Num rows: 1 Data size: 10 Basic stats: COMPLETE Column stats: NONE
          GatherStats: false
          Select Operator
            expressions: a (type: bigint), b (type: tinyint)
            outputColumnNames: _col0, _col1
            Statistics: Num rows: 1 Data size: 10 Basic stats: COMPLETE Column stats: NONE
            ListSink
```

**`expressions: a (type: bigint), b (type: tinyint)`**. I tried to alter the columns in the underlying tables to different types. I can see the types of view columns are always casted to the same one as the altered one

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: cc @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 2]
```diff
@@ -198,9 +203,44 @@ case class CatalogTable(
...
   /**
    * Return the default database name we use to resolve a view, should be None if the CatalogTable
-   * is not a View.
+   * is not a View or created by older versions of Spark(before 2.2.0).
+   */
+  def viewDefaultDatabase: Option[String] = properties.get(VIEW_DEFAULT_DATABASE)
+
+  /**
+   * Return the output column names of the query that creates a view, the column names are used to
+   * resolve a view, should be None if the CatalogTable is not a View or created by older versions
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: should be `Nil`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 3]
```diff
@@ -254,6 +294,9 @@ case class CatalogTable(
 
 object CatalogTable {
   val VIEW_DEFAULT_DATABASE = "view.default.database"
+  val VIEW_QUERY_OUTPUT_PREFIX = "view.query.out."
+  val VIEW_QUERY_OUTPUT_COLUMN_NUM = VIEW_QUERY_OUTPUT_PREFIX + "numCols"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: nit: `xxx_NUM_COLUMNS`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 4]
```diff
@@ -198,9 +203,44 @@ case class CatalogTable(
...
+  def viewDefaultDatabase: Option[String] = properties.get(VIEW_DEFAULT_DATABASE)
+
+  /**
+   * Return the output column names of the query that creates a view, the column names are used to
+   * resolve a view, should be None if the CatalogTable is not a View or created by older versions
+   * of Spark(before 2.2.0).
+   */
+  def viewQueryColumnNames: Seq[String] = {
+    for {
+      numCols <- properties.get(VIEW_QUERY_OUTPUT_COLUMN_NUM).toSeq
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: `.toSeq` is not needed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 5]
```diff
@@ -198,9 +203,44 @@ case class CatalogTable(
...
+        s"$numCols parts expected, but part $index is missing.")
+    )
+  }
+
+  /**
+   * Insert/Update the view query output column names in `properties`.
    */
-  def viewDefaultDatabase: Option[String] = properties.get(CatalogTable.VIEW_DEFAULT_DATABASE)
+  def withQueryColumnNames(columns: Seq[String]): CatalogTable = {
+    val props = new mutable.HashMap[String, String]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: let's follow the existing code for partition columns
```
properties.put(DATASOURCE_SCHEMA_NUMPARTCOLS, partitionColumns.length.toString)
      partitionColumns.zipWithIndex.foreach { case (partCol, index) =>
        properties.put(s"$DATASOURCE_SCHEMA_PARTCOL_PREFIX$index", partCol)
      }
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 6]
```diff
@@ -680,21 +700,70 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
     withTable("testTable") {
+      spark.range(1, 10).selectExpr("id", "id + 1 id1").write.saveAsTable("testTable")
       withView("testView") {
-        spark.range(1, 10).toDF("id1").write.format("json").saveAsTable("testTable")
-        sql("CREATE VIEW testView AS SELECT * FROM testTable")
+        val testView = CatalogTable(
+          identifier = TableIdentifier("testView"),
+          tableType = CatalogTableType.VIEW,
+          storage = CatalogStorageFormat.empty,
+          schema = new StructType().add("x", "long").add("y", "long"),
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: let's think about how we will persistent view when we have custom column names. Ideally we will have a logical plan representing the view, a SQL statement of the view query, and a `Seq[String]` for the custom column names.

1. call `plan.schema` to get the view schema, and zip it with custom column names, to get the final schema and save it. Then use `plan.schema.map(_.name)` to generate the `VIEW_QUERY_OUTPUT_COLUMN_NAME` in table properties.
2. call `plan.schema` to get the view schema, and save it as the final schema. Then use custom column names to generate the `VIEW_QUERY_OUTPUT_COLUMN_NAME` in table properties.

Personally I think option 2 is more natural, what do you think?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 7]
```diff
@@ -680,21 +700,70 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+          viewOriginalText = Some("SELECT * FROM testTable"),
+          viewText = Some("SELECT * FROM testTable"),
+          properties = Map(CatalogTable.VIEW_DEFAULT_DATABASE -> "default",
+            CatalogTable.VIEW_QUERY_OUTPUT_COLUMN_NUM -> "2",
+            s"${CatalogTable.VIEW_QUERY_OUTPUT_COLUMN_NAME_PREFIX}0" -> "id",
+            s"${CatalogTable.VIEW_QUERY_OUTPUT_COLUMN_NAME_PREFIX}1" -> "id1"))
+        hiveContext.sessionState.catalog.createTable(testView, ignoreIfExists = false)
+
+        // Correctly resolve a view with custom column names.
+        checkAnswer(sql("SELECT * FROM testView ORDER BY x"), (1 to 9).map(i => Row(i, i + 1)))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: can we use `select x, y ...`? to test the custom column names really work.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 8]
```diff
@@ -28,22 +28,56 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
-        val originAttr = findAttributeByName(attr.name, child.output, resolver)
-        // The dataType of the output attributes may be not the same with that of the view output,
-        // so we should cast the attribute to the dataType of the view output attribute. If the
-        // cast can't perform, will throw an AnalysisException.
-        Alias(Cast(originAttr, attr.dataType), attr.name)(exprId = attr.exprId,
-          qualifier = attr.qualifier, explicitMetadata = Some(attr.metadata))
+      val queryColumnNames = desc.viewQueryColumnNames
+      // If the view output doesn't have the same number of columns either with the child output,
+      // or with the query column names, throw an AnalysisException.
+      if (output.length != child.output.length && output.length != queryColumnNames.length) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: the comment says `or` but the code use `&&`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 4 above)
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: It is needed to generate the correct output.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 9]
```diff
@@ -28,22 +28,56 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
 /**
- * Make sure that a view's child plan produces the view's output attributes. We wrap the child
- * with a Project and add an alias for each output attribute. The attributes are resolved by
- * name. This should be only done after the batch of Resolution, because the view attributes are
- * not completely resolved during the batch of Resolution.
+ * Make sure that a view's child plan produces the view's output attributes. We try to wrap the
+ * child by:
+ * 1. Generate the `queryOutput` by:
+ *    1.1. If the query column names are defined, map the column names to attributes in the child
+ *         output by name;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: should we mention that, this is mostly for `SELECT * ...`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 10]
```diff
@@ -28,22 +28,56 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
-        val originAttr = findAttributeByName(attr.name, child.output, resolver)
-        // The dataType of the output attributes may be not the same with that of the view output,
-        // so we should cast the attribute to the dataType of the view output attribute. If the
-        // cast can't perform, will throw an AnalysisException.
-        Alias(Cast(originAttr, attr.dataType), attr.name)(exprId = attr.exprId,
-          qualifier = attr.qualifier, explicitMetadata = Some(attr.metadata))
+      val queryColumnNames = desc.viewQueryColumnNames
+      // If the view output doesn't have the same number of columns with the child output and the
+      // query column names, throw an AnalysisException.
+      if (output.length != child.output.length && output.length != queryColumnNames.length) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: This condition doesn't look very clear to me. How about `if (queryColumnNames.nonEmpty && output.length != queryColumnNames.length)`? When `queryColumnNames` is empty, it means this view is created prior to Spark 2.2, and we don't need to check anything.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 11]
```diff
@@ -28,22 +28,56 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
+        throw new AnalysisException(
+          s"The view output ${output.mkString("[", ",", "]")} doesn't have the same number of " +
+            s"columns with the child output ${child.output.mkString("[", ",", "]")}")
+      }
+      // If the child output is the same with the view output, we don't need to generate the query
+      // output again.
+      val queryOutput = if (queryColumnNames.nonEmpty && output != child.output) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: `output != child.output` will always be true right?

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: For a nested view, the inner view operator may have been resolved, in that case the output is the same with child.output.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 12]
```diff
@@ -28,22 +28,56 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
+        desc.viewQueryColumnNames.map { colName =>
+          findAttributeByName(colName, child.output, resolver)
+        }
+      } else {
+        child.output
+      }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: how about
```
val queryOutput = if (queryColumnNames.nonEmpty) {
  if (output.length != queryColumnNames.length) throw ...
  desc.viewQueryColumnNames.map { colName =>
    findAttributeByName(colName, child.output, resolver)
  }
} else {
  // For view created before Spark 2.1, the view text is already fully qualified, the plan output is view output.
  child.output
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala (hunk 11 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: shall we put this condition after the `case`? e.g. `case v @ View(desc, output, child) if child.resolved && output != child.output`

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: Oh I think that's better!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 13]
```diff
@@ -89,6 +89,25 @@ object Cast {
     case _ => false
   }
 
+  /**
+   * Return false iff we may truncate during casting `from` type to `to` type. e.g. long -> int,
+   * timestamp -> date.
+   */
+  def canUpCast(from: DataType, to: DataType): Boolean = (from, to) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: how about `def mayTruncate`? `canUpCast` is not accurate, we may not be able to cast even `canUpCast` returns true.

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala [hunk 14]
```diff
@@ -28,22 +28,60 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
+        desc.viewQueryColumnNames.map { colName =>
+          findAttributeByName(colName, child.output, resolver)
+        }
+      } else {
+        // For view created before Spark 2.2.0, the view text is already fully qualified, the plan
+        // output is the same with the view output.
+        child.output
+      }
+      // Map the attributes in the query output to the attributes in the view output by index.
+      val newOutput = output.zip(queryOutput).map {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: Seems we need to check the size of `output` and `queryOutput`.

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala]: For views created by older versions of Spark, the view text is fully qualified, so the output is the same with the view output. Or else we have checked that the output have the same length with `queryColumnNames`. So perhaps we don't need to check the size of `output` and `queryOutput` here.

**Comment by QQshu1** [comment]
hi , I have a question, why we should Eliminate View in the first of the optimizer.?
  thank you.@jiangxb1987

**Comment by jiangxb1987** [comment]
@QQshu1 As we have mentioned in the comment, the `View` operator is respected till the end of analysis stage to enable us better understand the analyzed logical plan. On the Beginning of optimize stage, that operator is no longer needed, so we apply the `EliminateView` rule.

**Comment by QQshu1** [comment]
thanks, What effect if we don`t Eliminate View？I  means whether it effect  optimize the tree namely performance or the  correctness of results ?

**Comment by cloud-fan** [comment]
the `View` operator doesn't have a corresponding physical operator, we have to remove it.

**Pull request closed without merging**